apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/ApolloHttpCache.kt

```diff
@@ -1,20 +1,31 @@
 package com.apollographql.apollo3.cache.http
 
+import com.apollographql.apollo3.annotations.ApolloDeprecatedSince
+import com.apollographql.apollo3.annotations.ApolloDeprecatedSince.Version.v3_2_3
 import com.apollographql.apollo3.api.http.HttpResponse
 import java.io.IOException
 
 interface ApolloHttpCache {
   fun read(cacheKey: String): HttpResponse
 
+  @Deprecated("Implementations should implement writeIncremental instead", ReplaceWith("writeIncremental(cacheKey, response)"))
+  @ApolloDeprecatedSince(v3_2_3)
+  fun write(response: HttpResponse, cacheKey: String)
+
   /**
    * Store the [response] with the given [cacheKey] into the cache.
-   * Note: the response's body is not consumed nor closed.
+   * A new [HttpResponse] is returned whose body, when read, will write the contents to the cache.
+   * The response's body is not consumed nor closed.
    */
-  fun write(response: HttpResponse, cacheKey: String)
+  fun writeIncremental(response: HttpResponse, cacheKey: String): HttpResponse {
```

Comment:
A bit unfortunate. Couldn't think of another way to do this.

Reply:
Is that even backward compatible? If someone implements `ApolloHttpCache` using `3.2.2` and ends up using `3.3.0` at runtime, what's going to happen? 

The Kotlin bytecode fir [CachingHttpInterceptor](https://github.com/apollographql/apollo-kotlin/blob/4b75f69669c26c0b3367eac0f50fab9dbd816a63/apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/CachingHttpInterceptor.kt#L113) shows 
```
    INVOKEVIRTUAL com/apollographql/apollo3/cache/http/DiskLruHttpCache.writeIncremental (Lcom/apollographql/apollo3/api/http/HttpResponse;Ljava/lang/String;)Lcom/apollographql/apollo3/api/http/HttpResponse;
```

Reply:
After looking a bit more at what we do with this interface:
- `CachingHttpInterceptor` instantiates `DiskLruHttpCache` directly, we don't provide a way to provide another implementation
- the `ApolloClient.httpCache: ApolloHttpCache` extension exposes the interface, but I think the only useful use-cases is to call `clearAll()` on it?

Could that mean that the interface is not really needed, and we could instead expose an extension `fun ApolloClient.clearHttpCache()`?

If yes we could:
- mark the whole interface as `@Deprecated`
- not implement it in `DiskLruHttpCache`
- mark `ApolloClient.httpCache` as deprecated, and make it return an implementation that only has `clearAll` working and the other ones being no-op

Reply:
> Could that mean that the interface is not really needed

I guess users could still call `write()` manually to warm the cache programmatically

> return an implementation that only has clearAll working and the other ones being no-op

That will make the above (hopefully very rare) use case fail so at this point, we might as well add the `HttpResponse` return value. The only thing that could break is if someone calls `ApolloHttpCache.write()` manually (very unlikely) and relies on the return value being `Unit` (even more unlikely)

Reply:
A lot simpler 😅  That makes sense 👍